h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -17,14 +17,47 @@
    *
    * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained. Only Frame and Model keys are accepted.
    */
-  public static void retain(final Key[] retainedKeys){
+  public static void retain(final Key... retainedKeys) {
     final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
     retainedSet.addAll(Arrays.asList(retainedKeys));
     // Frames and models have multiple nested keys. Those must be extracted and kept from deletion as well.
     extractNestedKeys(retainedSet);
 
-    new ClearDKVTask(retainedSet.toArray(new Key[retainedSet.size()]))
-            .doAllNodes();
+    for (final H2ONode node : H2O.CLOUD.members()) {
+      final NodeKeysRemovalTask nodeKeysRemovalTask = new NodeKeysRemovalTask(retainedSet.toArray(new Key[retainedSet.size()]));
```

Comment:
do you really need to do retainedSet.toArray every time?

Reply:
Of course not. Own code blindness.

Reply:
Should be resolved.